File: src/main/java/de/omilke/bankingfx/controls/DateEditingCell.java
package de.omilke.bankingfx.controls;

import javafx.scene.control.DatePicker;
import javafx.scene.control.TableCell;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public class DateEditingCell<P> extends TableCell<P, LocalDate> {

    private DatePicker datePicker;

    private final DateTimeFormatter formatter;

    public DateEditingCell(final DateTimeFormatter formatter) {

        this.formatter = formatter;
    }

    @Override
    public void startEdit() {

        if (!isEmpty()) {
            super.startEdit();
            createDatePicker();

            displayEditControl(getItem());

            this.datePicker.show();
        }
    }

    @Override
    public void cancelEdit() {

        super.cancelEdit();

        displayValue(getItem());
    }

    @Override
    public void updateItem(final LocalDate item, final boolean empty) {

        super.updateItem(item, empty);

        if (empty) {
            setText(null);
            setGraphic(null);
        } else {
            if (isEditing()) {

                displayEditControl(item);
            } else {

                displayValue(item);
            }
        }
    }

    private void displayValue(final LocalDate item) {

        setText(item.format(formatter));
        setGraphic(null);
    }

    private void displayEditControl(final LocalDate item) {

        if (datePicker != null) {
            datePicker.setValue(item);
        }

        setText(null);
        setGraphic(datePicker);
    }

    private void createDatePicker() {

        datePicker = new DatePicker(getItem());
        datePicker.setOnAction((e) -> commitEdit(this.datePicker.getValue()));
    }

}

File: src/main/java/de/omilke/bankingfx/controls/DateEditingCell.kt
package de.omilke.bankingfx.controls

import javafx.event.EventHandler
import javafx.scene.control.DatePicker
import javafx.scene.control.TableCell
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class DateEditingCell<P>(private val formatter: DateTimeFormatter) : TableCell<P, LocalDate?>() {

    //TODO: why lazily constructed?
    private var datePicker: DatePicker? = null

    override fun startEdit() {

        if (!isEmpty) {
            super.startEdit()
            createDatePicker()

            displayEditControl(item)

            datePicker!!.show()
        }
    }

    override fun cancelEdit() {

        super.cancelEdit()

        displayValue(item!!)
    }

    override fun updateItem(item: LocalDate?, empty: Boolean) {

        super.updateItem(item, empty)

        if (empty || item == null) {
            text = null
            setGraphic(null)
        } else {
            if (isEditing) {
                displayEditControl(item)
            } else {
                displayValue(item)
            }
        }
    }

    private fun displayValue(item: LocalDate) {

        text = item.format(formatter)
        graphic = null
    }

    private fun displayEditControl(item: LocalDate?) {

        datePicker?.let {
            it.value = item
        }

        text = null
        graphic = datePicker
    }

    private fun createDatePicker() {

        datePicker = DatePicker(item)
        datePicker!!.onAction = EventHandler { commitEdit(datePicker!!.value) }
    }
}